### PR TITLE
add ws2812 pio example

### PR DESCRIPTION
**Pull request by psnszsn**

### CODE REVIEW

File: src/hal/pio.zig [hunk 1]
```diff
@@ -480,7 +480,7 @@ pub const Pio = enum(u1) {
     ) !void {
         const expected_side_set_pins = if (program.side_set) |side_set|
             if (side_set.optional)
-                side_set.count - 1
+                side_set.count + 1
```
Comment (psnszsn) [inline src/hal/pio.zig:483]: If I understand the datasheet correctly this is supposed to be a `+`

Comment (mattnite) [reply src/hal/pio.zig:483]: I think we might be both incorrect here. The point of this check is to ensure that the user is mapping the number of pins equal to what they've defined in their program using the `.side_set` directive. The data sheet states:

> Additionally opt may be specified to indicate that a side <value> is optional for instructions (note this requires stealing an extra bit — in addition to the <count> bits

So in our case I don't think we need to check `side_set.optional` at all, we can make `expected_side_set_pins` equal to the count if it's set. Can you verify that on your example?

Thanks!

Comment (psnszsn) [reply src/hal/pio.zig:483]: From the datasheet:
> A side-set operation writes up to 5 bits. Depending on the register field EXECCTRL_SIDE_PINDIR, this is applied to either
> pins or pindirs. The least-significant bit of side-set data is mapped to PINCTRL_SIDESET_BASE, continuing for
> PINCTRL_SIDESET_COUNT pins, minus one if EXECCTRL_SIDE_EN is set.

EXECCTRL_SIDE_EN is set from `side_set.optional` in our case so if we want to use 1 pin with opt (`.side_set 1 opt`) we need to set PINCTRL_SIDESET_COUNT to 2.

In my example if I remove the assertion and use `.side_set 1 opt` , I have to set pin_mappings.side_set.count to 2. I tried with 1 and it does not work.

You can also see this behavior in the code generated by pioasm: for [uart_tx.pio](https://github.com/ZigEmbeddedGroup/raspberrypi-rp2040/blob/886234b88284bd90c951b930c98f28681140435b/src/hal/pio/assembler/comparison_tests/uart_tx.pio#L8) it calls  [`sm_config_set_sideset(&c, 2, true, false);`](https://github.com/ZigEmbeddedGroup/raspberrypi-rp2040/blob/886234b88284bd90c951b930c98f28681140435b/src/hal/pio/assembler/comparison_tests/uart_tx.pio.h#LL37C4-L37C47) in the generated code.

Comment (mattnite) [reply src/hal/pio.zig:483]: Okay thank you for explaining, I had the wrong idea in my head.

**Pull request merged**